upup/pkg/fi/cloudup/gce/gcediscovery/resolver.go

```diff
@@ -0,0 +1,266 @@
+func (r *Discovery) Region() string {
+	return r.region
+}
+
+// ClusterName returns the kOps cluster-name we are part of.
+func (r *Discovery) ClusterName() string {
+	return r.clusterName
+}
+
+// Compute returns the GCP compute service we built.
+func (r *Discovery) Compute() *compute.Service {
+	return r.compute
+}
+
+// New builds a Discovery.
+func New() (*Discovery, error) {
+	ctx := context.Background()
+
+	computeService, err := compute.NewService(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error building compute API client: %v", err)
+	}
+
+	myZoneName, err := metadata.Zone()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get zone from metadata: %w", err)
+	}
+
+	projectID, err := metadata.ProjectID()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get project id from metadata: %w", err)
+	}
+	projectID = strings.TrimSpace(projectID)
+
+	zones, err := computeService.Zones.List(projectID).Do()
+	if err != nil {
+		return nil, fmt.Errorf("error querying for GCE zones: %w", err)
+	}
+
+	// Find our zone
+	var myZone *compute.Zone
+	for _, zone := range zones.Items {
+		if myZoneName == zone.Name {
+			myZone = zone
+		}
+	}
+	if myZone == nil {
+		return nil, fmt.Errorf("failed to find zone %q", myZoneName)
+	}
+
+	region := lastComponent(myZone.Region)
+
+	// Find all the zones in our region
+	var zoneNames []string
+	for _, zone := range zones.Items {
+		regionName := lastComponent(zone.Region)
+		if regionName != region {
+			continue
+		}
+		zoneNames = append(zoneNames, zone.Name)
+	}
+
+	clusterName, err := metadata.InstanceAttributeValue(gcemetadata.MetadataKeyClusterName)
+	if err != nil {
+		return nil, fmt.Errorf("error reading cluster-name attribute from GCE: %w", err)
+	}
+	clusterName = strings.TrimSpace(clusterName)
+	if clusterName == "" {
+		return nil, fmt.Errorf("cluster-name metadata was empty")
+	}
+	klog.Infof("Found cluster-name=%q", clusterName)
+
+	return &Discovery{
+		compute:          computeService,
+		region:           region,
+		projectID:        projectID,
+		zone:             myZoneName,
+		clusterName:      clusterName,
+		allZonesInRegion: zoneNames,
+	}, nil
+}
+
+// Returns the last component of a URL, i.e. anything after the last slash
+// If there is no slash, returns the whole string
+func lastComponent(s string) string {
+	lastSlash := strings.LastIndex(s, "/")
+	if lastSlash != -1 {
+		s = s[lastSlash+1:]
+	}
+	return s
+}
+
+var _ resolver.Resolver = &Discovery{}
+
+// Resolve implements resolver.Resolve, providing name -> address resolution using GCE discovery.
+func (r *Discovery) Resolve(ctx context.Context, name string) ([]string, error) {
+	var records []string
+	klog.Infof("trying to resolve %q using GCEResolver", name)
+
+	var requiredTags []string
+
+	// We assume we are trying to resolve a component that runs on the control plane
+	requiredTags = append(requiredTags, gce.TagForRole(r.clusterName, kops.InstanceGroupRoleMaster))
```

Comment:
Should we also find instances tagged with InstanceGroupRoleAPIServer?

Reply:
I don't know; but here's why I think we don't need to do it yet..

1) We don't need _all_ the candidates, we just want to find one we can connect to
2) Do we run kops-controller on apiserver only nodes?  (And will we continue to do?)